### PR TITLE
smp server: combine messages and queue storage to optimise performance, prevent race condition when deleting queue and to avoid "orphan" messages.

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Messaging/Server/QueueStore/STM.hs [hunk 1]
```diff
@@ -1,120 +1,196 @@
...
+    $>>= \q -> q <$$ withLog "addQueue" st (`logCreateQueue` qr)
+  where
+    add = ifM hasId (pure $ Left DUPLICATE_) $ do
+      q <- mkQueue st qr -- STMQueue lock <$> (newTVar $! Just qr) <*> newTVar Nothing
+      TM.insert rId q $ queues' st
+      TM.insert sId rId $ senders' st
+      forM_ notifier $ \NtfCreds {notifierId} -> TM.insert notifierId rId $ notifiers' st
+      pure $ Right q
+    hasId = or <$> sequence [TM.member rId $ queues' st, TM.member sId $ senders' st, hasNotifier]
+    hasNotifier = maybe (pure False) (\NtfCreds {notifierId} -> TM.member notifierId (notifiers' st)) notifier
```
Comment (spaced4ndy) [inline src/Simplex/Messaging/Server/QueueStore/STM.hs:64]: wasn't checked before - was it a bug?

File: src/Simplex/Messaging/Server/QueueStore/STM.hs [hunk 2]
```diff
@@ -1,120 +1,196 @@
...
+updateQueueTime st sq t = atomically (readQueueRec qr >>= mapM update) $>>= log'
+  where
+    qr = queueRec' sq
+    update q@QueueRec {updatedAt}
+      | updatedAt == Just t = pure (q, False)
+      | otherwise =
+          let !q' = q {updatedAt = Just t}
+           in (writeTVar qr $! Just q') $> (q', True)
+    log' (q, changed)
+      | changed = q <$$ withLog "updateQueueTime" st (\sl -> logUpdateQueueTime sl (recipientId q) t)
```
Comment (spaced4ndy) [inline src/Simplex/Messaging/Server/QueueStore/STM.hs:134]: is it new logic - checking if time hasn't changed? does it affect anything?

File: src/Simplex/Messaging/Server/QueueStore/STM.hs (hunk 1 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Server/QueueStore/STM.hs:64]: We don't create queues with notifiers, it's only important when restoring. We do plan to create queues with user-supplied notifiers in the future, so added this check.

File: src/Simplex/Messaging/Server/QueueStore/STM.hs (hunk 2 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Server/QueueStore/STM.hs:134]: no, it prevents multiple writes with the same time in case of race condition.

Comment (epoberezkin) [reply src/Simplex/Messaging/Server/QueueStore/STM.hs:134]: we don't use this time anywhere yet.

File: src/Simplex/Messaging/Server/QueueStore/STM.hs [hunk 3]
```diff
@@ -1,120 +1,196 @@
...
+
+deleteQueue' :: STMQueueStore s => s -> RecipientId -> StoreQueue s -> IO (Either ErrorType (QueueRec, Maybe (MsgQueue s)))
+deleteQueue' st rId sq =
+  atomically (readQueueRec qr >>= mapM delete)
+    $>>= \q -> withLog "deleteQueue" st (`logDeleteQueue` rId)
+    >>= bimapM pure (\_ -> (q,) <$> atomically (swapTVar (msgQueue_' sq) Nothing))
+  where
+    qr = queueRec' sq
+    delete q = do
+      writeTVar qr Nothing
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply src/Simplex/Messaging/Server/QueueStore/STM.hs:145]: these comments are present in types

**Pull request merged**